src/aws_s3_presigned_url.erl

```diff
@@ -0,0 +1,150 @@
+%%%% @doc
+%%%
+%%% Allows generating either a get or put presigned s3 url.
+%%% This can be used by external clients such as cURL to access the object in question.
+%%%
+%%% See:
+%%%   - https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+%%%   - https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html
+-module(aws_s3_presigned_url).
+
+-export([ make_presigned_v4_url/5
+        ]).
+
+-include_lib("hackney/include/hackney_lib.hrl").
+
+%%====================================================================
+%% API
+%%====================================================================
+-spec make_presigned_v4_url(map(), get | put, integer(), binary(), binary()) -> {ok, binary()}.
+make_presigned_v4_url(Client0, Method, ExpireSeconds, Bucket, Key) ->
+    MethodBin = aws_request:method_to_binary(Method),
+    Path = ["/", aws_util:encode_uri(Bucket), "/", aws_util:encode_multi_segment_uri(Key), ""],
+    Client = Client0#{service => <<"s3">>},
+    SecurityToken = maps:get(token, Client, undefined),
+    AccessKeyID = maps:get(access_key_id, Client),
+    Region = maps:get(region, Client),
+    Host = build_host(<<"s3">>, Client, Bucket),
+    {BaseHost, URL0} = build_url(Host, Path, Client, Bucket),
+    Now = calendar:universal_time(),
+    ShortDate = list_to_binary(ec_date:format("Ymd", Now)),
+    Credential = make_credential(AccessKeyID, ShortDate, Region),
+    Headers0 = [{<<"Host">>, Host},
+                {<<"X-Amz-Algorithm">>, <<"AWS4-HMAC-SHA256">>},
+                {<<"X-Amz-Credential">>, Credential},
+                {<<"X-Amz-Date">>, list_to_binary(ec_date:format("YmdTHisZ", Now))},
+                {<<"X-Amz-Expires">>, integer_to_binary(ExpireSeconds)},
+                {<<"X-Amz-SignedHeaders">>, <<"host">>}
+               ],
+    Headers = maybe_add_security_token(Headers0, SecurityToken),
+    Signature = sign_request(Client, MethodBin, BaseHost, aws_request:add_query(URL0, Headers), [{<<"Host">>, BaseHost}], Now, <<"UNSIGNED-PAYLOAD">>),
+    FinalHeaders = lists:keysort(1, [{<<"X-Amz-Signature">>, Signature} | Headers]),
+    {ok, aws_request:add_query(URL0, FinalHeaders)}.
+
+%%====================================================================
+%% Internal functions
+%%====================================================================
+build_host(_EndpointPrefix, #{region := <<"local">>, endpoint := Endpoint}, undefined) ->
+    Endpoint;
+build_host(_EndpointPrefix, #{region := <<"local">>, endpoint := Endpoint}, _Bucket) ->
```

Comment:
This function is hard to read, and it seems like it can be clean up if the types are stricter.

Reply:
True, here we always have a Bucket so no need for the `_Bucket = undefined` clauses for starters. The remaining three function clauses will remain needed. Will push a touchup shortly.